staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go

```
@@ -0,0 +1,200 @@
+}
+
+// GetAPIServicesName returns the names of APIServices recorded in openAPIV3Specs.
+// We use this function to pass the names of local APIServices to the controller in this package,
+// so that the controller can periodically sync the OpenAPI spec from delegation API servers.
+func (s *specAggregator) GetAPIServiceNames() []string {
+	names := make([]string, len(s.openAPIV3Specs))
+	for key := range s.openAPIV3Specs {
+		names = append(names, key)
+	}
+	return names
+}
+
+// BuildAndRegisterAggregator registered OpenAPI aggregator handler. This function is not thread safe as it only being called on startup.
+func BuildAndRegisterAggregator(downloader Downloader, delegationTarget server.DelegationTarget, pathHandler common.PathHandlerByGroupVersion) (SpecAggregator, error) {
+	var err error
+	s := &specAggregator{
+		openAPIV3Specs: map[string]*openAPIV3APIServiceInfo{},
+		downloader:     downloader,
+	}
+
+	s.openAPIV3VersionedService, err = handler3.NewOpenAPIService(nil)
+	if err != nil {
+		return nil, err
+	}
+	err = s.openAPIV3VersionedService.RegisterOpenAPIV3VersionedService("/openapi/v3", pathHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	i := 1
+	for delegate := delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+		handler := delegate.UnprotectedHandler()
+		if handler == nil {
+			continue
+		}
+
+		apiServiceName := fmt.Sprintf(localDelegateChainNamePattern, i)
+		localAPIService := v1.APIService{}
+		localAPIService.Name = apiServiceName
+		s.AddUpdateAPIService(handler, &localAPIService)
+		s.UpdateAPIServiceSpec(apiServiceName)
+		i++
+	}
+
+	return s, nil
+}
+
+func (s *specAggregator) AddUpdateAPIService(handler http.Handler, apiservice *v1.APIService) error {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.openAPIV3Specs[apiservice.Name] = &openAPIV3APIServiceInfo{
+		apiService: *apiservice,
+		handler:    handler,
+		specs:      make(map[string]*openAPIV3SpecInfo),
+	}
+	return nil
```

does this need to return an error?

---

hmm looks like it doesn't. In removing an apiservice could have failed if the newly aggregated spec doesn't build properly, but with v3 we simply delete an entry in a map and it's impossible to have an error.

---

We can change the function signature then